File: December 2023/Check if a string is repetition of its substring of k-length/main.py
# Link: https://www.geeksforgeeks.org/problems/check-if-a-string-is-repetition-of-its-substring-of-k-length3302/1


class Solution:
    def kSubstrConcat(self, n, s, k):
        if n % k != 0:
            return 0
        unique_substrings, unique_substring_count = set(), 0
        for i in range(0, n, k):
            current_substring = s[i : i + k]
            if current_substring not in unique_substrings:
                if unique_substring_count == 2:
                    return 0
                else:
                    unique_substring_count += 1
                unique_substrings.add(current_substring)
        return 1


if __name__ == "__main__":
    obj = Solution()
    N = 4
    K = 2
    S = "bdac"
    print(obj.kSubstrConcat(N, S, K))

    N = 5
    K = 2
    S = "abcde"
    print(obj.kSubstrConcat(N, S, K))

File: December 2023/Check if a string is repetition of its substring of k-length/main.cpp
// Link: https://www.geeksforgeeks.org/problems/check-if-a-string-is-repetition-of-its-substring-of-k-length3302/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int kSubstrConcat(int n, string s, int k)
    {
        if (n % k || k == 0 || k == n)
            return 0;
        int division = n / k;
        unordered_map<string, int> mp;
        int count = 0;
        string temp = "";
        for (int i = 0; i < s.length(); i++)
        {
            count += 1;
            temp += s[i];
            if (count == k)
            {
                mp[temp]++;
                count = 0;
                temp = "";
            }
        }
        for (auto x : mp)
        {
            if (x.second == division)
                return 1;
            else if (x.second == division - 1)
                return 1;
        }
        return 0;
    }
};

int main()
{
    Solution obj;
    int N = 4;
    int K = 2;
    string S = "bdac";
    cout << obj.kSubstrConcat(N, S, K) << endl;

    N = 5;
    K = 2;
    S = "abcde";
    cout << obj.kSubstrConcat(N, S, K) << endl;

    return 0;
}